contrib/datadog/tracing/interceptor.go

```diff
@@ -0,0 +1,237 @@
+// that implements Datadog tracing for workflows.
+func NewInterceptor(opts InterceptorOptions) interceptor.Tracer {
+
+	return &tracerImpl{
+		opts: InterceptorOptions{
+			DisableSignalTracing: opts.DisableSignalTracing,
+			DisableQueryTracing:  opts.DisableQueryTracing,
+		},
+	}
+}
+
+type contextKey string
+
+const (
+	activeSpanContextKey contextKey = "dd_trace_span"
+	headerKey                       = string(activeSpanContextKey)
+)
+
+type tracerImpl struct {
+	interceptor.BaseTracer
+	// DisableSignalTracing can be set to disable signal tracing.
+	opts InterceptorOptions
+}
+
+func (t *tracerImpl) Options() interceptor.TracerOptions {
+	return interceptor.TracerOptions{
+		SpanContextKey:       activeSpanContextKey,
+		HeaderKey:            headerKey,
+		DisableSignalTracing: t.opts.DisableSignalTracing,
+		DisableQueryTracing:  t.opts.DisableQueryTracing,
+	}
+}
+
+func (t *tracerImpl) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	var carrier tracer.TextMapCarrier = m
+	ctx, err := tracer.Extract(carrier)
+	if err != nil {
+		if errors.Is(err, tracer.ErrSpanContextNotFound) {
+			// If there is no span, return nothing, but don't error out. This is
+			// a legitimate place where a span does not exist in the headers
+			return nil, nil
+		}
+		return nil, err
+	}
+	return &tracerSpanCtx{ctx}, nil
+}
+
+func (t *tracerImpl) MarshalSpan(span interceptor.TracerSpan) (map[string]string, error) {
+	carrier := tracer.TextMapCarrier{}
+	if err := tracer.Inject(span.(*tracerSpan).Context(), carrier); err != nil {
+		return nil, err
+	}
+	return carrier, nil
+}
+
+func (t *tracerImpl) SpanFromContext(ctx context.Context) interceptor.TracerSpan {
+	span, ok := tracer.SpanFromContext(ctx)
+	if !ok {
+		return nil
+	}
+	return &tracerSpan{Span: span}
+}
+
+func (t *tracerImpl) ContextWithSpan(ctx context.Context, span interceptor.TracerSpan) context.Context {
+	return tracer.ContextWithSpan(ctx, span.(*tracerSpan).Span)
+}
+
+func genSpanID(idempotencyKey string) uint64 {
+	h := fnv.New64()
+	// Write() always writes all bytes and never fails; the count and error result are for implementing io.Writer.
+	_, _ = h.Write([]byte(idempotencyKey))
+	return h.Sum64()
+}
+
+func (t *tracerImpl) StartSpan(options *interceptor.TracerStartSpanOptions) (interceptor.TracerSpan, error) {
+	startOpts := []tracer.StartSpanOption{
+		tracer.ResourceName(options.Name),
+		tracer.StartTime(options.Time),
+	}
+	// Set a deterministic span ID for workflows which are long-running and cross process boundaries
+	if options.Operation == "RunWorkflow" {
+		startOpts = append(startOpts, tracer.WithSpanID(genSpanID(options.IdempotencyKey)))
+	}
+
+	// Add parent span to start options
+	var parent ddtrace.SpanContext
+	switch opParent := options.Parent.(type) {
+	case nil:
+	case *tracerSpan:
+		parent = opParent.Context()
+	case *tracerSpanCtx:
+		parent = opParent
+	default:
+		return nil, fmt.Errorf("unrecognized parent type %T", opParent)
```

Comment:
I think this is because of how the datadog library handles a parent trace depending if it comes from network or it comes from memory (somewhere else in the same code).
You are absolutely right for the default case, we should just log it and have a `nil` parent in that case.